services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java

```diff
@@ -0,0 +1,726 @@
+    public String updateAll(@ApiParam(value = "The number of messages to send per second", required = true) @RequestParam long sendRate,
+                    HttpServletResponse response) {
+        
+        log.info("Updating sendRate to " + sendRate + " for all audit replays");
+        
+        String resp;
+        
+        // only update if the send rate is valid
+        if (sendRate >= 0) {
+            resp = updateAll(sendRate, replayProperties.isPublishEvents(), false) + "audit replays updated";
+        } else {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            resp = "Send rate must be >= 0";
+        }
+        
+        log.info(resp);
+        return resp;
+    }
+    
+    private int updateAll(long sendRate, boolean publishEnabled, boolean onlyUpdateRunning) {
+        int replaysUpdated = 0;
+        
+        if (publishEnabled)
+            appCtx.publishEvent(new AuditReplayRemoteRequestEvent(this, busProperties.getId(), Request.updateAll(sendRate)));
+        
+        // update all of our replays.
+        List<Status> statuses = statusAll();
+        if (onlyUpdateRunning)
+            statuses = statuses.stream().filter(status -> status.getState() == RUNNING).collect(Collectors.toList());
+        
+        for (Status status : statuses) {
+            update(status, sendRate, false);
+            replaysUpdated++;
+        }
+        
+        return replaysUpdated;
+    }
+    
+    /**
+     * Stops an audit replay
+     *
+     * @param id
+     *            The audit replay id
+     * @return status, indicating whether the audit replay was successfully stopped
+     */
+    @ApiOperation(value = "Stops an audit replay.")
+    @RolesAllowed({"Administrator", "JBossAdministrator"})
+    @RequestMapping(path = "/{id}/stop", method = org.springframework.web.bind.annotation.RequestMethod.POST)
+    public String stop(@ApiParam("The audit replay id") @PathVariable("id") String id, HttpServletResponse response) {
+        
+        log.info("Stopping audit replay with id " + id);
+        
+        String resp;
+        
+        Status status = status(id, replayProperties.isPublishEvents());
+        if (status != null) {
+            if (stop(status, replayProperties.isPublishEvents())) {
+                resp = "Stopped audit replay with id " + id;
+            } else {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                resp = "Cannot stop audit replay with id " + id;
+            }
+        } else {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            resp = "No audit replay found with id " + id;
+        }
+        
+        log.info(resp);
+        return resp;
+    }
+    
+    private boolean stop(Status status, boolean publishEvent) {
+        // is the replay running?
+        if (status.getState() == RUNNING) {
```

Comment:
This is a lot to parse, but I'm wondering if these status object can be updated by multiple threads? If so, then you need to synchronize here, yes?

Reply:
I thought about this a bit...  A status could be updated by multiple threads, or even by multiple audit services depending upon which pod you get when you hit the audit endpoint.  

To protect against race conditions I'll need some kind of distributed lock.  From the looks of it, Hazelcast has locking functionality built in.  I suppose I could either update the existing HazelcastCacheInspector, or extend it and add some methods which expose the locking functionality.  Thoughts?

Reply:
I'll add that given how I envision the replay endpoints being used, I doubt we would run encounter race condition issues with the code as-is, but I don't think it will be hard to utilize the Hazelcast locking functionality, and that would solve the problem.

Reply:
Ok, so the next commit will update the restcontroller to acquire a lock on the Hazelcast map before updating the status of a replay.  For running replays, updates and stop commands are forwarded to the running replay object, which will be responsible for updating the cache.

Reply:
this looks good